File: kafka/src/test/groovy/io/micronaut/configuration/kafka/docs/consumer/offsets/manual/ProductListener.java
package io.micronaut.configuration.kafka.docs.consumer.offsets.manual;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.configuration.kafka.docs.consumer.config.Product;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.OffsetAndMetadata;
import org.apache.kafka.common.TopicPartition;

import java.util.Collections;

import static io.micronaut.configuration.kafka.annotation.OffsetReset.EARLIEST;
import static io.micronaut.configuration.kafka.annotation.OffsetStrategy.DISABLED;
// end::imports[]

class ProductListener {

    // tag::method[]
    @KafkaListener(offsetReset = EARLIEST, offsetStrategy = DISABLED) // <1>
    @Topic("awesome-products")
    void receive(Product product,
                 long offset,
                 int partition,
                 String topic,
                 Consumer kafkaConsumer) { // <2>
        // process product record

        // commit offsets
        kafkaConsumer.commitSync(Collections.singletonMap( // <3>
                new TopicPartition(topic, partition),
                new OffsetAndMetadata(offset + 1, "my metadata")
        ));
    }
    // end::method[]
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/offsets/manual/ProductListener.py
# tag::imports[]
from java.lang import Long
from java.util import Collections
from micronaut.configuration.kafka.annotation import KafkaListener, OffsetReset, OffsetStrategy, Topic
from micronaut.context.annotation import Requires
from micronaut.kafka.docs.Product import Product
from org.apache.kafka.clients.consumer import Consumer, OffsetAndMetadata
from org.apache.kafka.common import TopicPartition
# end::imports[]


@Requires(property="spec.name", value="ManualProductListenerTest")
# tag::clazz[]
@KafkaListener(offsetReset=OffsetReset.EARLIEST, offsetStrategy=OffsetStrategy.DISABLED)  # <1>
class ProductListener:

    @Topic("awesome-products")
    def receive(self, product: Product, offset: Long, partition: int, topic: str, kafka_consumer: Consumer) -> None:  # <2>
        # process product record
        # commit offsets
        kafka_consumer.commitSync(Collections.singletonMap(  # <3>
            TopicPartition(topic, partition),
            OffsetAndMetadata(offset + 1, "my metadata")
        ))
# end::clazz[]
